### ISSUE TITLE
MacOS build error (member function needExactTags)

### ISSUE THREAD

**Issue by FlugelderFreiheit** [Bionus/imgbrd-grabber#1316]
### What steps will reproduce the problem?
Building the app in MacOS both using the build-mac.sh script and manually fails. make fails with following error.
4 errors generated.
make[2]: *** [tests/CMakeFiles/tests.dir/src/models/filename-test.cpp.o] Error 1
make[1]: *** [tests/CMakeFiles/tests.dir/all] Error 2
make: *** [all] Error 2

### What is the expected behavior? What do you get instead?
App is expected to be built completely but fails before make is complete.

### How often does this problem occur? On which sources?
I was able to build from source on version 5.5.2 but since version 6.0.0 and above make fails.

### What version of the program are you using? On what operating system?
MacOS 10.13.5
Grabber v6.0.3
Compiler used is clang APPLE LLVM 9.1.0
Changing CMAKE_C_COMPILER and CMAKE_CXX_COMPILER to gcc-8 and g++-8 gives error of 
'init_priority' attribute is not supported on this platform during make.

### Please provide any additional information below
The error during make is as below.

/Users/username/Downloads/imgbrd-grabber-master/tests/src/models/filename-test.cpp:673:35: error: 
      no matching member function for call to 'needExactTags'
        QCOMPARE(Filename("%md5%.%ext%").needExactTags(false), 0);
                 ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/usr/local/opt/qt5/lib/QtTest.framework/Headers/qtestcase.h:87:26: note: 
      expanded from macro 'QCOMPARE'
    if (!QTest::qCompare(actual, expected, #actual, #expected, __FILE__...
                         ^~~~~~
/Users/username/Downloads/imgbrd-grabber-master/tests/../lib/src/models/filename.h:46:7: note: 
      candidate function not viable: no known conversion from 'bool' to 'Site *'
      for 1st argument
                int needExactTags(Site *site, const QString &api = "") const;
                    ^
/Users/username/Downloads/imgbrd-grabber-master/tests/../lib/src/models/filename.h:47:7: note: 
      candidate function not viable: no known conversion from 'bool' to
      'const QStringList' for 1st argument
                int needExactTags(const QStringList &forcedTokens = QStr...
                    ^
/Users/username/Downloads/imgbrd-grabber-master/tests/src/models/filename-test.cpp:675:50: error: 
      no matching member function for call to 'needExactTags'
        QCOMPARE(Filename("javascript:md5 + '.' + ext").needExactTags(fa...
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/usr/local/opt/qt5/lib/QtTest.framework/Headers/qtestcase.h:87:26: note: 
      expanded from macro 'QCOMPARE'
    if (!QTest::qCompare(actual, expected, #actual, #expected, __FILE__...
                         ^~~~~~
/Users/username/Downloads/imgbrd-grabber-master/tests/../lib/src/models/filename.h:46:7: note: 
      candidate function not viable: no known conversion from 'bool' to 'Site *'
      for 1st argument
                int needExactTags(Site *site, const QString &api = "") const;
                    ^
/Users/username/Downloads/imgbrd-grabber-master/tests/../lib/src/models/filename.h:47:7: note: 
      candidate function not viable: no known conversion from 'bool' to
      'const QStringList' for 1st argument
                int needExactTags(const QStringList &forcedTokens = QStr...
                    ^
/Users/username/Downloads/imgbrd-grabber-master/tests/src/models/filename-test.cpp:676:47: error: 
      no matching member function for call to 'needExactTags'
        QCOMPARE(Filename("%character% %md5%.%ext%").needExactTags(false), 1);
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/usr/local/opt/qt5/lib/QtTest.framework/Headers/qtestcase.h:87:26: note: 
      expanded from macro 'QCOMPARE'
    if (!QTest::qCompare(actual, expected, #actual, #expected, __FILE__...
                         ^~~~~~
/Users/username/Downloads/imgbrd-grabber-master/tests/../lib/src/models/filename.h:46:7: note: 
      candidate function not viable: no known conversion from 'bool' to 'Site *'
      for 1st argument
                int needExactTags(Site *site, const QString &api = "") const;
                    ^
/Users/username/Downloads/imgbrd-grabber-master/tests/../lib/src/models/filename.h:47:7: note: 
      candidate function not viable: no known conversion from 'bool' to
      'const QStringList' for 1st argument
                int needExactTags(const QStringList &forcedTokens = QStr...
                    ^
/Users/username/Downloads/imgbrd-grabber-master/tests/src/models/filename-test.cpp:677:58: error: 
      no matching member function for call to 'needExactTags'
        QCOMPARE(Filename("%all:includenamespace% %md5%.%ext%").needExactTags(fa...
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/usr/local/opt/qt5/lib/QtTest.framework/Headers/qtestcase.h:87:26: note: 
      expanded from macro 'QCOMPARE'
    if (!QTest::qCompare(actual, expected, #actual, #expected, __FILE__...
                         ^~~~~~
/Users/username/Downloads/imgbrd-grabber-master/tests/../lib/src/models/filename.h:46:7: note: 
      candidate function not viable: no known conversion from 'bool' to 'Site *'
      for 1st argument
                int needExactTags(Site *site, const QString &api = "") const;
                    ^
/Users/username/Downloads/imgbrd-grabber-master/tests/../lib/src/models/filename.h:47:7: note: 
      candidate function not viable: no known conversion from 'bool' to
      'const QStringList' for 1st argument
                int needExactTags(const QStringList &forcedTokens = QStr...
                    ^
4 errors generated.
make[2]: *** [tests/CMakeFiles/tests.dir/src/models/filename-test.cpp.o] Error 1
make[1]: *** [tests/CMakeFiles/tests.dir/all] Error 2
make: *** [all] Error 2

**Comment by UntouchedWagons** [comment]
I'm pretty sure I'm having the same issue on Fedora 28

**Comment by chrissly90** [comment]
I've git the same Problem on Debian 9
Notice, that's my first attempt to troubleshooting code on Github, The solution may not be perfect.
 
The Funktion it's complaining about is INT (see gui/src/mainwindow.h)
`int needExactTags(QSettings *settings);`

But the Funktion is called with a Boolean. I simply replaced the "false" with a "0"
tests/src/models/filename-test.cpp Line 646
`
void FilenameTest::testNeedExactTags()
{
        //QCOMPARE(Filename("%md5%.%ext%").needExactTags(false), 0);
        QCOMPARE(Filename("%md5%.%ext%").needExactTags(0), 0);
        QCOMPARE(Filename("%md5%.%ext%").needExactTags(m_site), 0);
        //QCOMPARE(Filename("javascript:md5 + '.' + ext").needExactTags(false), 2);
        QCOMPARE(Filename("javascript:md5 + '.' + ext").needExactTags(0), 2);
        //QCOMPARE(Filename("%character% %md5%.%ext%").needExactTags(false), 1);
        QCOMPARE(Filename("%character% %md5%.%ext%").needExactTags(0), 1);
        //QCOMPARE(Filename("%all:includenamespace% %md5%.%ext%").needExactTags(false), 1);
        QCOMPARE(Filename("%all:includenamespace% %md5%.%ext%").needExactTags(0), 1);

        Filename filename("%filename%.%ext%");
        QCOMPARE(filename.needExactTags(), 0);
        QCOMPARE(filename.needExactTags(QStringList() << "filename"), 2);

        Filename date("%date%.%ext%");
        QCOMPARE(date.needExactTags(), 0);
        QCOMPARE(date.needExactTags(QStringList() << "date"), 2);
}
`
Now it's compiling successfully

**Comment by FlugelderFreiheit** [comment]
Thanks @chrissly90 That fixed it. But I think the right thing to do would be to replace 'false' with NULL instead of 0 since the function call is expecting a pointer or object reference as its first argument and not int. This will help prevent future issues of the same kind. This issue is probably because older version of gcc which was used to test the code supported implicit conversion of boolean false to NULL pointer but the newer versions of gcc and clang do not support this for type safety. Maybe you can try building the unmodified code using older version of gcc such as gcc-4.9 or gcc-5 and see if you are getting the same error. I am not able to build on MacOS using gcc so maybe you can try this.

I will leave this issue open so that someone can take a look and modify the code in the next release.

**Comment by chrissly90** [comment]
I just compiled again, just this time I replaced "0" with "NULL". The build was successfull, too.
For gcc-4.9 or gcc-5 I would need to set up a VM, maybe someone else can test it?

**Comment by FlugelderFreiheit** [comment]
So I tried compiling on my bootable ubuntu. First I installed all versions of gcc 4.8.5, 5.5.0, 6.4.0, 7.3.0 and 8.1.0. Next I tried building setting -DCMAKE_C_COMPILER flag and -DCMAKE_CXX_COMPILER flags to each version. The compiling was successful for gcc-4.8 and gcc-5 though there is a warning mentioned for converting 'false' to pointer type. For versions equal to or above gcc-6 and clang the implicit conversion from boolean to pointer type throws a error and compilation fails.
So anyone who doesn't want to modify the source files can just install gcc-5 and use the above cmake flags to use gcc-5 and g++-5 for compilation.
